### PR TITLE
shorten compliant username

### PR DESCRIPTION
**Pull request by ranakan19**
fix wrt https://issues.redhat.com/browse/CRT-1670

Continue to keep the transformed username to be within the maxLength of 20 characters when dealing with forbidden suffixes and preffixes in generateCompliantUsername. 
shortening the compliant username impacts the MUR name and other related resources. Changes in unit tests and e2e tests for the same.

related PRs:
https://github.com/codeready-toolchain/toolchain-common/pull/309
https://github.com/codeready-toolchain/toolchain-e2e/pull/714
https://github.com/codeready-toolchain/registration-service/pull/320

### CODE REVIEW

**Comment by ranakan19** [comment]
e2e updated
/retest

**Comment by ranakan19** [comment]
maybe a flaky test,
/retest

**Comment by ranakan19** [comment]
updated e2e
/retest

File: controllers/usersignup/usersignup_controller.go [hunk 1]
```diff
@@ -512,20 +512,32 @@ func (r *Reconciler) updateUserSignupMetricsByState(oldState string, newState st
...
-
+	// -4 for "crt-" or "-crt" to be added in following lines. Update this length changing prefix
+	maxlengthWithPrefix := usersignup.MaxLength - 4
 	// Check for any forbidden prefixes
 	for _, prefix := range config.Users().ForbiddenUsernamePrefixes() {
 		if strings.HasPrefix(replaced, prefix) {
-			replaced = fmt.Sprintf("%s%s", "crt-", replaced)
+			if len(replaced) > maxlengthWithPrefix {
+				// replace prefix instead of append
+				replaced = "crt-" + replaced[3:]
```
Comment (alexeykazakov) [inline controllers/usersignup/usersignup_controller.go:524]: Maybe we should move the forbidden username checks and replacements to the `usersignup.TransformUsername()` function? So we have a single place for all username transformations.

File: controllers/usersignup/usersignup_controller.go [hunk 2]
```diff
@@ -512,20 +512,32 @@ func (r *Reconciler) updateUserSignupMetricsByState(oldState string, newState st
 }
 
 func (r *Reconciler) generateCompliantUsername(config toolchainconfig.ToolchainConfig, instance *toolchainv1alpha1.UserSignup) (string, error) {
+	// replaced should now be of len
```
Comment (rajivnathan) [inline controllers/usersignup/usersignup_controller.go:515]: Incomplete comment?

File: controllers/usersignup/usersignup_controller.go (hunk 1 above)
Comment (rajivnathan) [reply controllers/usersignup/usersignup_controller.go:524]: +1

**Review by rajivnathan — COMMENTED** [review]
Looks good overall!

**Review by sbryzak — APPROVED** [review]

File: controllers/usersignup/usersignup_controller.go (hunk 1 above)
Comment (ranakan19) [reply controllers/usersignup/usersignup_controller.go:524]: sure we can definitely move prefix and suffix checks.
But not QualifiedName check, an unqualified name will block UserSignup creation - which I don't think we want to do.

Comment (ranakan19) [reply controllers/usersignup/usersignup_controller.go]: done in [047f883](https://github.com/codeready-toolchain/host-operator/pull/794/commits/047f883d16ec5b6326c2d88f295239f994112669)
Also removed the qualifiedName check in [b7cb183](https://github.com/codeready-toolchain/host-operator/pull/794/commits/b7cb18362ad0d592fcc97a1a7e57ae7d78b43df1)as we check for DNSLabel check in toolchain-common

File: controllers/usersignup/usersignup_controller.go (hunk 2 above)
Comment (ranakan19) [reply controllers/usersignup/usersignup_controller.go]: fixed in [c14186a](https://github.com/codeready-toolchain/host-operator/pull/794/commits/c14186aa4f94ab6a9da5bfdb44a3549c4c17256d)

**Review by alexeykazakov — APPROVED** [review]

**Comment by ranakan19** [comment]
> 9c5c57d

@MatousJobanek added a check in here - [9c5c57d](https://github.com/codeready-toolchain/host-operator/pull/794/commits/9c5c57d58a3b52130eca97b30b88148789c71c3f)

**Comment by ranakan19** [comment]
> Looks good 👍 correct me if I'm wrong - we actually don't verify the length of the compliantUsername anywhere in the unit tests, right? If not, then could you please add that there?

@MatousJobanek added a check in here - [9c5c57d](https://github.com/codeready-toolchain/host-operator/pull/794/commits/9c5c57d58a3b52130eca97b30b88148789c71c3f)

**Pull request merged**